src/number_theory/primorial.lean

```diff
@@ -0,0 +1,190 @@
+than or equal to `n`.
+-/
+
+open finset
+open nat
+open_locale big_operators
+
+/-- The primorial `n#` of `n` is the product of the primes less than or equal to `n`.
+-/
+def primorial (n : ℕ) : ℕ := ∏ p in (filter prime (range (n + 1))), p
+local notation x`#` := primorial x
+
+lemma primorial_succ {n : ℕ} (n_big : 1 < n) (r : n % 2 = 1) : (n + 1)# = n# :=
+begin
+  have not_prime : ¬prime (n + 1),
+    { intros is_prime,
+      cases (prime.eq_two_or_odd is_prime),
+      { linarith, },
+      { exfalso,
+        rw ←not_even_iff at h r,
+        have e : even (n + 1 - n), exact (even_sub (le_of_lt (lt_add_one n))).2 (iff_of_false h r),
+        simp only [nat.add_sub_cancel_left, not_even_one] at e,
+        exact e, }, },
+  apply finset.prod_congr,
+  { rw [@range_succ (n + 1), filter_insert, if_neg not_prime], },
+  { exact λ _ _, rfl, },
+end
+
+private lemma range_contains_upper_bound (n : ℕ) : n ∈ range (n + 1) :=
+by simp only [succ_pos', lt_add_iff_pos_right, mem_range]
+
+private lemma select_from_sum (f : ℕ → ℕ) (s : finset ℕ) (a : ℕ) (t : a ∈ s) : f a ≤ ∑ i in s, f i :=
+by rw <- finset.insert_erase t; simp
+
+private lemma bound_binomial_coefficient (n : ℕ) : choose (2 * n + 1) n ≤ 4 ^ n :=
+begin
+  have t : choose (2 * n + 1) n ≤ ∑ i in range (n + 1), choose (2 * n + 1) i,
+    exact select_from_sum (choose (2 * n + 1)) (range (n + 1)) n (range_contains_upper_bound n),
+  simpa [sum_range_choose_halfway n] using t
+end
+
+lemma dvd_choose_of_middling_prime (p : ℕ) (is_prime : prime p) (m : ℕ)
+  (p_big : m + 1 < p) (p_small : p ≤ 2 * m + 1) : p ∣ choose (2 * m + 1) (m + 1) :=
+begin
+  have m_size : m + 1 ≤ 2 * m + 1, by exact le_of_lt (lt_of_lt_of_le p_big p_small),
+  let e := @choose_mul_fact_mul_fact (2 * m + 1) (m + 1) m_size,
+  have r : p ∣ nat.fact (2 * m + 1), exact (prime.dvd_fact is_prime).mpr p_small,
+  have s : (p ∣ nat.fact (m + 1)) → false,
+    { intros,
+      let e := (prime.dvd_fact is_prime).mp a,
+      linarith, },
+  have t : (p ∣ nat.fact (2 * m + 1 - (m + 1))) → false,
+    { intros,
+      let f := (prime.dvd_fact is_prime).mp a,
+      have t : 2 * m + 1 - (m + 1) = m, { norm_num, rw two_mul m, exact nat.add_sub_cancel m m, },
+      rw t at f,
+      clear t a s r p_small is_prime e m_size,
+      cases lt_trichotomy p m,
+      { have r : m < m + 1, exact lt_add_one m, linarith, },
+      { cases h,
+        { rw h at p_big,
+          linarith, },
+        linarith, }, },
+  rw [←e, mul_assoc] at r,
+  cases (prime.dvd_mul is_prime).1 r,
+  { exact h, },
+  cases (prime.dvd_mul is_prime).1 h,
+  cc, cc,
+end
+
+lemma prod_primes_dvd {s : finset ℕ} : ∀ (n : ℕ) (h : ∀ a ∈ s, prime a) (div : ∀ a ∈ s, a ∣ n),
+  (∏ p in s, p) ∣ n :=
+begin
+  apply finset.induction_on s,
+  { simp, },
+  { intros a s a_not_in_s induct n primes divs,
+    rw finset.prod_insert a_not_in_s,
+    have a_div_n : a ∣ n, by exact divs a (finset.mem_insert_self a s),
```

Comment:
```suggestion
    obtain ⟨k, rfl⟩ : a ∣ n, by exact divs a (finset.mem_insert_self a s),
```
this way you automatically do `rcases` on the result, and the equality in the second hypothesis is immediately substituted.